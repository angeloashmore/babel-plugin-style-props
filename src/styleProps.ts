import { types as t } from '@babel/core'
import { JSXAttribute, ObjectProperty } from '@babel/types'
import { buildObjectProperty } from 'builders'
import { PluginOptions, StylePropExpression } from '../types'
import { STYLE_ALIASES } from './constants'
import { castArray, shouldSkipProp } from './utils'

const processProp = (
  cssPropertyNames: string[],
  expression: StylePropExpression,
  result: ObjectProperty[]
) => {
  cssPropertyNames.forEach(cssName => {
    if (shouldSkipProp(expression)) return

    result.push(buildObjectProperty(cssName, expression!))
  })
}

export const processStyleProps = (
  options: PluginOptions,
  styleProps: JSXAttribute[]
) => {
  const baseResult = [] as ObjectProperty[]
  const responsiveResults = [] as ObjectProperty[][]

  styleProps.forEach(prop => {
    const { variants } = options
    const propName = prop.name.name as string
    const propValue = prop.value
    const cssPropertyNames = castArray(STYLE_ALIASES[propName] || propName)

    if (t.isJSXExpressionContainer(propValue)) {
      const expression = propValue.expression as StylePropExpression

      if (t.isArrayExpression(expression)) {
        // e.g. prop={['foo', null, 'bar']}
        const elements = expression.elements as StylePropExpression[]
        elements.forEach((element, i) => {
          let resultObj = baseResult

          if (i !== 0) {
            responsiveResults[i - 1] = responsiveResults[i - 1] || []
            resultObj = responsiveResults[i - 1]
          }

          processProp(cssPropertyNames, element, resultObj)
        })
      } else {
        // e.g. prop={text} prop={bool ? 'foo' : 'bar'}
        processProp(cssPropertyNames, expression, baseResult)
      }
    } else {
      // e.g. prop="test"
      const isVariant = Boolean(variants[propName])
      if (isVariant) {
        // TODO
      }

      processProp(cssPropertyNames, propValue, baseResult)
    }
  })

  const baseResultObj = t.objectExpression(baseResult)
  const responsiveResultObjs = responsiveResults.map(properties =>
    t.objectExpression(properties)
  )

  const resultArr = t.arrayExpression([baseResultObj, ...responsiveResultObjs])

  return buildObjectProperty('base', resultArr)
}
