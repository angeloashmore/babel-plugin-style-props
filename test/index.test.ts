import pluginTester from 'babel-plugin-tester'
import babelConfig from '../babel.config'
import styleProps from '../src'

pluginTester({
  plugin: styleProps,
  babelOptions: babelConfig,
  snapshot: true,

  tests: {
    'runs without dying': {
      code: `
        const Box = () => {
          const boolean = false

          return <div mx={['3rem', null, '4rem']} lineHeight={boolean} />
        }
      `,
    },
  },
})
