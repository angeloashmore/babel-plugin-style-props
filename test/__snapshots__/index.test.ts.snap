// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kitchen sink handles a large amount of scale and style props 1`] = `
"const Example = () => {
  const array = ['l', 'l', 'm', 'm', 'xl'];
  const variable = 'huge';
  return <div display='flex' mScale={array} fontSize={['1rem', '2rem', null, '3rem']} color='green' lineHeight={1.5} pyScale={['l', 'l', 'xxl', 'xxl', 'xxl']} textTransform='uppercase' fontFamily='system-ui' maxWidth={variable} __styleProps__={[{
    display: 'flex',
    fontSize: '1rem',
    color: 'green',
    lineHeight: 1.5,
    textTransform: 'uppercase',
    fontFamily: 'system-ui',
    maxWidth: variable,
    margin: array[0],
    paddingTop: 'l'[0],
    paddingBottom: 'l'[0]
  }, {
    fontSize: '2rem',
    margin: array[1],
    paddingTop: 'l'[1],
    paddingBottom: 'l'[1]
  }, {
    margin: array[2],
    paddingTop: 'xxl'[2],
    paddingBottom: 'xxl'[2]
  }, {
    fontSize: '3rem',
    margin: array[3],
    paddingTop: 'xxl'[3],
    paddingBottom: 'xxl'[3]
  }, {
    margin: array[4],
    paddingTop: 'xxl'[4],
    paddingBottom: 'xxl'[4]
  }]} />;
};"
`;

exports[`scale prop parsing handles responsive scale props 1`] = `
"const Example = () => {
  return <div mScale={['l', 'l', 'm', 'm', 'm']} __styleProps__={[{
    margin: 'l'[0]
  }, {
    margin: 'l'[1]
  }, {
    margin: 'm'[2]
  }, {
    margin: 'm'[3]
  }, {
    margin: 'm'[4]
  }]} />;
};"
`;

exports[`scale prop parsing handles scale props 1`] = `
"const Example = () => {
  return <div mScale='l' __styleProps__={[{
    margin: 'l'[0]
  }, {
    margin: 'l'[1]
  }, {
    margin: 'l'[2]
  }, {
    margin: 'l'[3]
  }, {
    margin: 'l'[4]
  }]} />;
};"
`;

exports[`scale prop parsing handles variable arrays in scale props 1`] = `
"const Example = () => {
  const array = ['l', 'l', 'm', 'm', 'xl'];
  return <div mScale={array} __styleProps__={[{
    margin: array[0]
  }, {
    margin: array[1]
  }, {
    margin: array[2]
  }, {
    margin: array[3]
  }, {
    margin: array[4]
  }]} />;
};"
`;

exports[`scale prop parsing merges scale props with an existing __styleProps__ prop 1`] = `
"const Example = () => {
  return <div pScale='l' __styleProps__={[{
    margin: '1rem',
    padding: 'l'[0]
  }, {
    padding: 'l'[1]
  }, {
    margin: '3rem',
    padding: 'l'[2]
  }, {
    margin: '4rem',
    padding: 'l'[3]
  }, {
    padding: 'l'[4]
  }]} />;
};"
`;

exports[`style prop parsing does not strip non style props if \`shouldStrip\` is set 1`] = `
"const Example = () => {
  return <div shouldNotStrip={true} __styleProps__={[{
    padding: '1rem'
  }, {
    padding: '2rem'
  }, {
    padding: '3rem'
  }, {
    padding: '4rem'
  }]} />;
};"
`;

exports[`style prop parsing handles expressions in style props 1`] = `
"const Example = ({
  isTest
}) => {
  return <div m={isTest ? '3rem' : '4rem'} __styleProps__={[{
    margin: isTest ? '3rem' : '4rem'
  }]} />;
};"
`;

exports[`style prop parsing handles responsive style props 1`] = `
"const Example = () => {
  return <div m={['3rem', '4rem']} display='grid' pt={[null, '4rem', null, '6rem']} __styleProps__={[{
    margin: '3rem',
    display: 'grid'
  }, {
    margin: '4rem',
    paddingTop: '4rem'
  }, {}, {
    paddingTop: '6rem'
  }]} />;
};"
`;

exports[`style prop parsing handles style props and places them in a new prop 1`] = `
"const Example = () => {
  return <div m='3rem' lineHeight={1.5} __styleProps__={[{
    margin: '3rem',
    lineHeight: 1.5
  }]} />;
};"
`;

exports[`style prop parsing handles style props on multiple elements 1`] = `
"const Example = () => {
  return <div m='1rem' __styleProps__={[{
    margin: '1rem'
  }]}>
            <span p='2rem' __styleProps__={[{
      padding: '2rem'
    }]} />
          </div>;
};"
`;

exports[`style prop parsing handles variable usage in style props 1`] = `
"const Example = ({
  size
}) => {
  const variable = '3rem';

  const myFunction = () => '4rem';

  return <div m={[variable, size, myFunction()]} __styleProps__={[{
    margin: variable
  }, {
    margin: size
  }, {
    margin: myFunction()
  }]} />;
};"
`;

exports[`style prop parsing merges parsed props with an existing __styleProps__ prop 1`] = `
"const Example = () => {
  return <div p={['1rem', '2rem', '3rem', '4rem']} __styleProps__={[{
    margin: '1rem',
    padding: '1rem'
  }, {
    padding: '2rem'
  }, {
    margin: '3rem',
    padding: '3rem'
  }, {
    margin: '4rem',
    padding: '4rem'
  }]} />;
};"
`;

exports[`style prop parsing strips style props if \`shouldStrip\` is set 1`] = `
"const Example = () => {
  return <div __styleProps__={[{
    padding: '1rem'
  }, {
    padding: '2rem'
  }, {
    padding: '3rem'
  }, {
    padding: '4rem'
  }]} />;
};"
`;
